File: services/feed-ingester/internal/oddinxml/specifiers.go
// Market specifier canonicalization.
//
// INVARIANT: the output of Canonical() and Hash() MUST match the TypeScript
// implementation at packages/types/src/specifiers.ts byte-for-byte. Both are
// tested against docs/fixtures/specifiers.json. If this diverges, settlement
// worker silently fails to match tickets to settled markets.

package oddinxml

import (
	"crypto/sha256"
	"sort"
	"strings"
)

// Specifiers is the parsed (key, value) map from Oddin's pipe-separated form.
type Specifiers map[string]string

// Parse converts a raw Oddin specifiers attribute (`k1=v1|k2=v2`) into a map.
// Values containing `=` are preserved (split only at the first `=`).
// Returns nil for an empty input.
func Parse(raw string) Specifiers {
	if raw == "" {
		return Specifiers{}
	}
	out := make(Specifiers)
	for _, pair := range strings.Split(raw, "|") {
		eq := strings.IndexByte(pair, '=')
		if eq < 0 {
			continue
		}
		out[pair[:eq]] = pair[eq+1:]
	}
	return out
}

// Canonical returns the canonical pipe-separated form: `k1=v1|k2=v2` with
// keys sorted lexicographically by byte. An empty map serializes to "".
func Canonical(s Specifiers) string {
	if len(s) == 0 {
		return ""
	}
	keys := make([]string, 0, len(s))
	for k := range s {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	var buf strings.Builder
	for i, k := range keys {
		if i > 0 {
			buf.WriteByte('|')
		}
		buf.WriteString(k)
		buf.WriteByte('=')
		buf.WriteString(s[k])
	}
	return buf.String()
}

// Hash returns sha256 of the canonical form as a 32-byte slice, suitable
// for storing in markets.specifiers_hash (BYTEA).
func Hash(s Specifiers) []byte {
	sum := sha256.Sum256([]byte(Canonical(s)))
	return sum[:]
}

// HashFromRaw is a convenience for when the caller has the raw attribute.
func HashFromRaw(raw string) []byte {
	return Hash(Parse(raw))
}

File: packages/types/src/specifiers.ts
// Canonical market-specifier representation. The output of `canonical()` is
// what feed-ingester stores in markets.specifiers_json. `hash()` produces the
// 32-byte sha256 stored in markets.specifiers_hash.
//
// CRITICAL: the Go implementation at
// services/feed-ingester/internal/oddinxml/specifiers.go must produce the
// byte-identical canonical string and hash for the same input.

import { createHash } from "node:crypto";

export type Specifiers = Record<string, string>;

/** Canonical pipe-separated form: `k1=v1|k2=v2`, keys sorted lexicographically. */
export function canonical(specs: Specifiers): string {
  const keys = Object.keys(specs).sort();
  return keys.map((k) => `${k}=${specs[k]}`).join("|");
}

/** 32-byte sha256 of the canonical form, as a Buffer. */
export function hash(specs: Specifiers): Buffer {
  return createHash("sha256").update(canonical(specs), "utf8").digest();
}

/** Parse an Oddin-style "k=v|k=v" string back into a Specifiers map. */
export function parse(raw: string): Specifiers {
  if (!raw) return {};
  const out: Specifiers = {};
  for (const pair of raw.split("|")) {
    const eq = pair.indexOf("=");
    if (eq < 0) continue;
    out[pair.slice(0, eq)] = pair.slice(eq + 1);
  }
  return out;
}
